### ISSUE TITLE
Feature: Implement Modular Model Management System for Agent Actions (Move, Attack, etc.)

### ISSUE THREAD

**Issue by csmangum** [Dooders/Experiments#19]
Develop a modularized system for managing different models required for various agent actions, such as moving, attacking, and defending. This system should allow for clean, maintainable, and scalable code by organizing each action’s model and logic into separate modules. The agent will then interact with these modules, enabling it to select actions and perform them based on its state and health. 

This modular structure will simplify adding new actions in the future, enhance the readability of the code, and support flexible and independent model training.

---

### **Detailed Requirements**

1. **Modularize Agent Actions into Separate Classes**
   - **Move Module**:
     - Create a `MoveModule` class that contains a Q-network (`MoveQNetwork`) for moving the agent in various directions (e.g., up, down, left, right).
     - Define a training method for updating the Q-network based on experience (e.g., rewards for successful moves).
     - Implement an `action_space` attribute to define available move actions.
   - **Attack Module**:
     - Create an `AttackModule` class with a Q-network (`AttackQNetwork`) for handling attack actions in different directions and defending.
     - Define a health-based rule within the module to boost the probability of the defend action when the agent’s health is below a threshold (e.g., 30%).
     - Define a training method for updating the attack model based on experience.
     - Implement an `action_space` attribute to define available attack actions.

2. **Agent Class Integration**
   - **Composition of Action Modules**:
     - Modify the `Agent` class to include instances of `MoveModule` and `AttackModule`.
     - Allow the agent to select an action by checking its state and health, delegating action selection to the appropriate module.
   - **Centralized Decision-Making**:
     - Implement a `decide_action` method in the `Agent` class to determine the appropriate action (move or attack) based on context.
     - Implement a `should_attack` method or similar logic to decide whether the agent should perform an attack or a move.

3. **Model Encapsulation within Modules**
   - Each module (e.g., `MoveModule`, `AttackModule`) should encapsulate its Q-network model, optimizer, and training logic to promote separation of concerns.
   - Use separate optimizers and loss functions for each module to allow for independent training of models.

4. **Model Loading and Saving**
   - Implement `save_model` and `load_model` methods in each module to handle model checkpointing and persistence.
   - Update the `Agent` class with `save_models` and `load_models` methods that call each module’s save and load functions, saving all models to a specified filepath.

5. **Unified Training Management**
   - Implement a training method in the `Agent` class that directs the training to the correct module based on the action taken.
   - **Training Logic**:
     - Each module’s training method should calculate target Q-values based on rewards and the gamma parameter.
     - Use Mean Squared Error (MSE) as the loss function for each Q-network model.
   - Optional: Use experience replay within modules to improve training stability by sampling batches from stored experiences.

6. **Experience Replay (Optional Enhancement)**
   - Consider implementing a shared experience replay buffer for each module to improve training efficiency and prevent overfitting to recent experiences.

7. **Testing and Validation**
   - After implementing the modular structure, create tests to validate:
     - The agent can correctly load, save, and retrieve model states for both the move and attack modules.
     - Each module correctly trains its Q-network, updating weights and improving action selection over time.
     - The agent’s `decide_action` method can consistently select appropriate actions based on different states and health levels.
     - Rule-based constraints (e.g., health threshold for defending) work as expected, preventing the agent from taking unnecessary risks when health is low.

8. **Code Structure and Organization**
   - Organize code into a clear directory structure for maintainability:
     ```
     /agents
       /modules
         move_module.py
         attack_module.py
       agent.py
     ```
   - Create documentation for each module to explain its purpose, key methods, and usage.

---

### **Acceptance Criteria**

- [x] **Action Modules**: Each action module (move and attack) encapsulates its own model, optimizer, and training logic, and is properly integrated with the `Agent` class.
- [x] **Decision-Making**: The `Agent` class can accurately determine which action to take and delegate action selection to the appropriate module based on state and health.
- [x] **Model Checkpointing**: Models can be saved and loaded individually, or collectively by the `Agent` class, ensuring persistence across training sessions.
- [x] **Training**: Each module’s training process independently updates its Q-network, demonstrating improvements over random or heuristic-based approaches.
- [x] **Testing**: All functionality is covered by tests, including action selection, model persistence, rule-based constraints, and training behavior.

---

### **Additional Context**

This modular system should streamline development, making it easier to add new actions or modify existing ones without extensive changes. The agent’s ability to balance aggressive and defensive actions will improve over time, benefiting from both learned behaviors and predefined survival rules.

**Labels**: `enhancement`, `modularization`, `Q-learning`, `agent-behavior`, `model-management`

**Comment by csmangum** [comment]
### **1. Modularize Agent Actions**

**Action Modules**:

- **Create Separate Modules**: For each action (e.g., move, attack), create separate classes or modules that handle the specifics of that action, including its neural network model and associated logic.

  ```python
  class MoveModule:
      def __init__(self):
          self.model = MoveQNetwork()
          self.optimizer = optim.Adam(self.model.parameters(), lr=0.001)
      # Move-specific methods...

  class AttackModule:
      def __init__(self):
          self.model = AttackQNetwork()
          self.optimizer = optim.Adam(self.model.parameters(), lr=0.001)
      # Attack-specific methods...
  ```

**Agent Class Integration**:

- **Compose Modules**: Your `Agent` class will then compose these modules.

  ```python
  class Agent:
      def __init__(self):
          self.move_module = MoveModule()
          self.attack_module = AttackModule()
          # Other attributes like health, position, etc.
  ```

---

### **2. Manage Models within Each Module**

**Model Encapsulation**:

- **Encapsulate Models**: Keep the neural network and its training logic within the module. This makes the code cleaner and each module self-contained.

**Separate Optimizers and Training Loops**:

- **Optimizers**: Use separate optimizers for each model to allow independent training.

  ```python
  self.optimizer = optim.Adam(self.model.parameters(), lr=0.001)
  ```

- **Training Methods**: Each module has its own training method.

  ```python
  def train(self, state, action, reward, next_state, done):
      # Training logic specific to this action
      pass
  ```

---

### **3. Action Decision Handling**

**Centralized Decision Method**:

- **Decide Action**: Implement a method in the `Agent` class to decide which action to take and delegate to the appropriate module.

  ```python
  def decide_action(self, state):
      if self.should_attack(state):
          return self.attack_module.select_action(state, self.health)
      else:
          return self.move_module.select_action(state)
  ```

**Condition for Action Selection**:

- **Logic**: Define the conditions under which the agent decides to move or attack, possibly based on state evaluation or other criteria.

  ```python
  def should_attack(self, state):
      # Logic to determine if the agent should attack
      pass
  ```

---

### **4. Model Loading and Saving**

**Checkpointing Models**:

- **Save Models**:

  ```python
  def save_models(self, filepath):
      torch.save({
          'move_model_state_dict': self.move_module.model.state_dict(),
          'attack_model_state_dict': self.attack_module.model.state_dict(),
      }, filepath)
  ```

- **Load Models**:

  ```python
  def load_models(self, filepath):
      checkpoint = torch.load(filepath)
      self.move_module.model.load_state_dict(checkpoint['move_model_state_dict'])
      self.attack_module.model.load_state_dict(checkpoint['attack_model_state_dict'])
  ```

---

### **5. Shared vs. Individual Models**

**Shared Models**:

- **Advantages**: Saves memory and allows collective learning.
- **Implementation**: Pass shared model instances to each agent.

  ```python
  shared_attack_model = AttackQNetwork()
  # Load pre-trained weights if necessary

  # In Agent initialization
  self.attack_module = AttackModule(model=shared_attack_model)
  ```

**Individual Models**:

- **Advantages**: Allows for agent-specific behaviors and learning.
- **Implementation**: Each agent instantiates its own models.

---

### **6. Training Management**

**Unified Training Loop**:

- **Agent-Level Training**: The `Agent` class can have a training method that delegates training to the appropriate module based on the action taken.

  ```python
  def train(self, state, action, reward, next_state, done):
      if action in self.attack_module.action_space:
          self.attack_module.train(state, action, reward, next_state, done)
      elif action in self.move_module.action_space:
          self.move_module.train(state, action, reward, next_state, done)
  ```

**Experience Replay (Optional)**:

- **Shared Replay Buffer**: For shared models, consider using a shared experience replay buffer to improve learning efficiency.

---

### **7. Structuring Your Codebase**

**Directory Organization**:

- **Modules in Separate Files**:

  ```
  /agents
    /modules
      move_module.py
      attack_module.py
    agent.py
  ```

**Configuration Management**:

- **Use Config Files**: Externalize hyperparameters and settings using configuration files (YAML, JSON).

---

### **8. Model Registry and Dynamic Loading**

**Model Registry**:

- **Purpose**: Helps manage multiple models, especially if you have various types of actions or agents.

  ```python
  class ModelRegistry:
      def __init__(self):
          self.registry = {}

      def register(self, action_name, model):
          self.registry[action_name] = model

      def get_model(self, action_name):
          return self.registry.get(action_name)
  ```

---

### **9. Memory and Performance Optimization**

**Lazy Initialization**:

- **Load on Demand**: Initialize models only when required.

**Shared Memory Spaces**:

- **Efficiency**: Use shared memory to avoid duplicate models when agents share models.

---

### **10. Logging, Monitoring, and Testing**

**Logging**:

- **Per-Module Logging**: Implement logging within each module for better traceability.

**Monitoring**:

- **Visualization Tools**: Use TensorBoard or similar tools to monitor training progress.

**Testing**:

- **Unit Tests**: Write tests for each module to ensure they're working as expected.

---

### **11. Frameworks and Libraries**

**Leverage RL Libraries**:

- **Stable Baselines3**, **Ray RLlib**, or **TensorFlow Agents** can help manage complex architectures and offer utilities for multi-agent systems.

---

### **12. Example Implementation**

Here's an example combining these concepts:

```python
# move_module.py
class MoveModule:
    def __init__(self, model=None):
        self.model = model or MoveQNetwork()
        self.optimizer = optim.Adam(self.model.parameters(), lr=0.001)
        self.action_space = ['move_up', 'move_down', 'move_left', 'move_right']

    def select_action(self, state):
        # Logic to select move action
        pass

    def train(self, state, action, reward, next_state, done):
        # Training logic
        pass

# attack_module.py
class AttackModule:
    def __init__(self, model=None):
        self.model = model or AttackQNetwork()
        self.optimizer = optim.Adam(self.model.parameters(), lr=0.001)
        self.action_space = ['attack_up', 'attack_down', 'attack_left', 'attack_right', 'defend']

    def select_action(self, state, health):
        # Logic to select attack action
        pass

    def train(self, state, action, reward, next_state, done):
        # Training logic
        pass

# agent.py
class Agent:
    def __init__(self, move_module=None, attack_module=None):
        self.move_module = move_module or MoveModule()
        self.attack_module = attack_module or AttackModule()
        self.health = 1.0  # Example health value

    def decide_action(self, state):
        if self.should_attack(state):
            action = self.attack_module.select_action(state, self.health)
        else:
            action = self.move_module.select_action(state)
        return action

    def should_attack(self, state):
        # Logic to decide whether to attack
        return True  # Placeholder

    def train(self, state, action, reward, next_state, done):
        if action in self.attack_module.action_space:
            self.attack_module.train(state, action, reward, next_state, done)
        elif action in self.move_module.action_space:
            self.move_module.train(state, action, reward, next_state, done)

    def save_models(self, filepath):
        self.move_module.save_model(filepath + '_move.pth')
        self.attack_module.save_model(filepath + '_attack.pth')

    def load_models(self, filepath):
        self.move_module.load_model(filepath + '_move.pth')
        self.attack_module.load_model(filepath + '_attack.pth')
```

---

### **13. Additional Considerations**

**Scalability**:

- **Adding New Actions**: With this structure, adding a new action involves creating a new module and integrating it into the `Agent` class.

**Collaboration**:

- **Team Development**: Modular code allows different team members to work on separate modules simultaneously.

**Maintenance**:

- **Isolation**: Bugs or changes in one module are less likely to impact others.

---

### **14. Advanced Techniques**

**Policy Networks vs. Q-Networks**:

- Depending on your simulation, you might consider using policy gradient methods or actor-critic models, which can also be modularized similarly.

**Meta-Controller**:

- Implement a higher-level controller that decides which action module should be active based on the agent's goals.

---

### **15. Conclusion**

By modularizing your agent's actions and encapsulating models within dedicated modules, you create a clean, maintainable, and scalable codebase. This approach:

- **Simplifies Management**: Each model is managed within its own context.
- **Enhances Flexibility**: Easily add or modify actions without significant restructuring.
- **Improves Readability**: Clear separation of concerns makes the code easier to understand.
